models/multiphase/d3q27_pf_velocity/Dynamics.c.Rt

```diff
@@ -0,0 +1,714 @@
+//			    Code builds and runs stable for density
+//			    ratio up to 100, but layered Poiseuille
+//			    flow does not appear symmetric,#DebugFunTimes
+//			    - #DebugFunTimes have passed, validated with:
+//				- Layered Poiseuille flow
+//				- Rayleigh Taylor up to water-air like properties and Re=3000
+//				- Validated with air Taylor bubble rising through olive oil
+//				  experiment by Bugg et al. (2002)
+//		14/08/2017: Look to improve readability by incorporating more R code
+//				- e.g. for MRT operations etc.
+
+<?R
+    source("conf.R")
+    c_header();
+    g=PV(DensityAll$name[DensityAll$group=="g"])
+# Extracting velocity set
+    U = as.matrix(DensityAll[DensityAll$group=="g",c("dx","dy","dz")])
+    u = PV(c("U","V","W"))
+    absCi2 = (U[,1]*U[,1]+U[,2]*U[,2]+U[,3]*U[,3])
+
+# Formulate Weighted-MRT matrix
+    m0 = matrix(1, 1, 27)
+    m1 = U[,1]
+    m2 = U[,2]
+    m3 = U[,3]
+    m4 = U[,1]*U[,2]
+    m5 = U[,2]*U[,3]
+    m6 = U[,3]*U[,1]
+    m7 = 3*U[,1]*U[,1] - absCi2
+    m8 = U[,2]*U[,2] - U[,3]*U[,3]
+    m9 = absCi2 - 1
+    m10 = U[,1] * (3*absCi2 - 5)
+    m11 = U[,2] * (3*absCi2 - 5)
+    m12 = U[,3] * (3*absCi2 - 5)
+    m13 = U[,1] * (U[,2]*U[,2] - U[,3]*U[,3])
+    m14 = U[,2] * (U[,3]*U[,3] - U[,1]*U[,1])
+    m15 = U[,3] * (U[,1]*U[,1] - U[,2]*U[,2])
+    m16 = U[,1]*U[,2]*U[,3]
+    m17 = 0.5 * (3*absCi2*absCi2 - 9*absCi2+4)
+    m18 = (3*absCi2 - 4) * (3*U[,1]*U[,1] - absCi2)
+    m19 = (3*absCi2 - 4) * (3*U[,2]*U[,2] - U[,3]*U[,3])
+    m20 = U[,1]*U[,2]*(3*absCi2 - 7)
+    m21 = U[,2]*U[,3]*(3*absCi2 - 7)
+    m22 = U[,3]*U[,1]*(3*absCi2 - 7)
+    m23 = 0.5 * U[,1]*(absCi2*absCi2 - absCi2 + 2)
+    m24 = 0.5 * U[,2]*(absCi2*absCi2 - absCi2 + 2)
+    m25 = 0.5 * U[,3]*(absCi2*absCi2 - absCi2 + 2)
+    m26 = 0.5 * (9*absCi2*absCi2*absCi2 - 36*absCi2*absCi2 + 33*absCi2 - 2)
+
+M = rbind(m0,m1,m2,m3,m4,m5,m6,m7,m8,m9,m10,m11,m12,m13,m14,m15,m16,m17,m18,m19,m20,m21,m22,m23,m24,m25,m26)
+m = PV(c(paste("m[",1:27-1,"]",sep="")))
+invM = solve(M)
+
+geq = PV(c(paste("geq[",1:27-1,"]",sep="")))
+Fi  = PV(c(paste("F_i[",1:27-1,"]",sep="")))
+h   = PV(c(paste("h",1:15-1,sep="")))
+omega = PV("omega_phi")
+heq  = PV(c(paste("heq[",1:15-1,"]",sep="")))
+Fphi = PV(c(paste("F_phi[",1:15-1,"]",sep="")))
+w_h = c(0.75, 1.5, 1.5, 1.5, 1.5, 1.5, 1.5, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0)
+
+phase = PV("C")
+gammah = PV(c(paste("Gamma[",1:15-1,"]",sep="")))
+gammag = PV(c(paste("Gamma[",1:27-1,"]",sep="")))
+?>
+
+#include <math.h>
+#define PI 3.14159265
+
+//	EXPORTABLE QUANTITIES:
+CudaDeviceFunction real_t getPhaseField(){
+	return PhaseF(0,0,0);
+}
+CudaDeviceFunction vector_t getU(){
+	vector_t u;
+	u.x = U;
+	u.y = V;
+	u.z = W;
+	return u;
+}
+CudaDeviceFunction real_t getP(){
+	real_t   p = <?R C(sum(g)) ?>;
+	return p;
+}
+
+// 	HELPER FUNCTIONS:
+CudaDeviceFunction vector_t calcGradPhi(){
+	vector_t gradPhi = {0.0,0.0,0.0};
+	if (NodeType & NODE_SymmetricXY_W) {
```

Comment:
If you have multiple nodetypes in one group it is a bad idea to use "&" for checking. It should be:
```c
if ((NodeType & NODE_ADDITIONALS) == NODE_SymmetricXY_W) {
```